﻿First draft of gate change and settings

RyanM-RMA opened the pull request:
Includes DTO and DAO implementations, and unit and integration tests respectively

RyanM-RMA committed:
﻿First draft of gate change and settings
Includes DTO and DAO implementations, and unit and integration tests respectively

RyanM-RMA committed:
Removing oracle type map references

RyanM-RMA committed:
Adjusting logic for change IT
loops over the collection and finds the right field, then compares those, instead of expecting index based comparisons to work.

RyanM-RMA committed:
Adjusting the zone id to America/Los_Angeles
This comes from the TurbineChangeTest

rma-bryson commented on cwms-data-api/src/main/java/cwms/cda/data/dao/location/kind/OutletDao.java:
```diff
@@ -242,4 +253,116 @@ public void renameOutlet(String officeId, String oldOutletId, String newOutletId
+                    formatBool(startInclusive), formatBool(endInclusive), rowLimitBig);
+
+            List<GateChange> output = new ArrayList<>();
+            if (changeTab != null) {
```
If changeTab can come back null, does that mean no gate changes were found? If so, I would throw a NotFoundException

RyanM-RMA replied:
Threw a NotFoundException, that's a good idea, I was going off what TurbineDao did, but I think this is a better approach.

rma-bryson commented on cwms-data-api/src/test/java/cwms/cda/data/dto/location/kind/GateChangeTest.java:
Add test of missing required fields

RyanM-RMA replied:
Added test for this

rma-bryson commented on cwms-data-api/src/main/java/cwms/cda/data/dto/location/kind/PhysicalStructureChange.java:
```diff
@@ -56,7 +56,6 @@ public abstract class PhysicalStructureChange<T extends Setting> extends CwmsDTO
     private final Double newTotalDischargeOverride;
     private final Double oldTotalDischargeOverride;
     private final String dischargeUnits;
-    private final Double poolElevation;
```
I believe you could just keep the poolElevation in this base class. Then in the class that requires it (GateChange), you can override the validator

`@Override
    protected void validateInternal(CwmsDTOValidator validator)
    {
        validator.required(getPoolElevation(), "poolElevation");
        super.validateInternal(validator);
    }`

RyanM-RMA replied:
Validation isn't really the reason I did any of this.  The public API for the two change classes is what I was considering when making this change.  poolElevation is required for gate changes, but not for turbines, so for turbines the method `public Double getPoolElevation()` makes sense.  Whereas this doesn't make sense for the gate changes.  We can't override the `getPoolElevation()` to return a double instead, which is why the API is split apart, and since poolElevation is a private field it's also separated with the public API.
If you can think of a better approach to keeping that API that way, let me know and I'd be glad to consider that.

RyanM-RMA replied:
I updated it to use Double's, as we had discussed, and had to adjust the GateSetting to do the same.  I like the idea of the validation handling this.

rma-bryson commented on cwms-data-api/src/test/java/cwms/cda/data/dto/location/kind/GateSettingTest.java:
Add missing required fields test

RyanM-RMA replied:
Added test for this

rma-bryson commented on cwms-data-api/src/main/java/cwms/cda/data/dto/location/kind/GateChange.java:
```diff
@@ -0,0 +1,87 @@
+
+    GateChange(Builder builder) {
+        super(builder);
+        this.poolElevation = Objects.requireNonNull(builder.poolElevation, "Pool elevation is required when creating a GateChange");
```
Remove this Objects.requireNonNull. Null checks should get handled by the validation process on required fields.

RyanM-RMA replied:
Removed

RyanM-RMA committed:
Throwing NotFoundException instead of null check

RyanM-RMA committed:
Updating GateChange and Setting to use Double
Added missing field validation tests, and updated the code to use Double objects, instead of primitives

rma-psmorris commented on cwms-data-api/src/test/resources/cwms/cda/data/dto/location/kind/gate-settings.json:
```diff
@@ -0,0 +1,24 @@
+[
+  {
+    "type": "gate-setting",
```
this data structure needs to scale similar to time series

RyanM-RMA replied:
Do you have some time today or this week where we can talk about this?
